prediction_market_agent/agents/microchain_agent/microchain_agent.py

```diff
@@ -80,8 +83,22 @@ def build_agent(
         engine.register(f)
 
     agent.max_tries = 3
+
+    # We restore the prompt from a historical session.
+    if prompt_handler:
+        if historical_prompt := prompt_handler.fetch_latest_prompt():
+            system_prompt = historical_prompt.prompt
+
     print(system_prompt)
-    agent.system_prompt = system_prompt.format(engine_help=engine.help)
+    # if {engine_help} not in prompt, we expect the functions to have been already loaded,
+    # thus no need to load them again. Otherwise we can simply not use the historical prompt.
+    agent.system_prompt = system_prompt
+    if "{engine_help}" not in system_prompt:
```

Comment:
This feels quite hacky. This is something I didn't pick up in https://github.com/gnosis/prediction-market-agent/pull/222, but I think `build_agent` shouldn't take a system prompt template as an arg, it should take the fully-formatted system prompt. Then we don't need to make any assumptions about what the prompt contains here.

The way I'd do that is by moving the engine initialization into a separate function, `build_agent_engine`, and then `build_agent` takes an engine arg, and `system_prompt` is the actual prompt. wdyt?

Reply:
Agree it's a bit hacky and like the overall idea, however not sure how to handle the fact that this assignment (L95)
```
agent.system_prompt = system_prompt
```
is needed for agent initialization. Hence we need the system_prompt here still.

My counter-proposal would be to move the hacky bit to a separate function (`maybe_append_functions_into_prompt` or similar), and that gets called during agent initialization.
Bear in mind it's nothing magical, it simply adds the function definitions to the prompt being initialized in case it is still not there. Wdyt?

Reply:
See https://github.com/gnosis/prediction-market-agent/commit/d2d892136b033d71255e18ba62b2860bb527fc1d#diff-59ec0cef505f2b78726896b047967f06a472aeb8b7bb5ec33baab7e43fff720aR57 for the current suggestion

Reply:
> however not sure how to handle the fact that this assignment ...

Not sure I understand. That could just stay where it is?

So before:
```
def build_agent(
    system_prompt: str,
    ....
):
    engine = Engine()
    agent = Agent(engine, ...)
    agent.system_prompt = system_prompt.format(engine_help=engine.help)
```

After:

```
engine = ...

def build_agent(
    system_prompt: str,
    engine: Engine,
    ....
):
    agent = Agent(engine)
    agent.system_prompt = system_prompt
```

But my point is that sequence of operations should be:

- Build the engine
- Build the system prompt, either by formatting `TRADING_AGENT_SYSTEM_PROMPT` or by fetching a prompt from the db (not by looking for a particular string in the prompt)
- Build the agent.

With `maybe_append_function_definitions_into_prompt` the hackyness is just moved somewhere else.

Reply:
Created this issue for tracking this discussion. Let's make this part of a separate PR (agree with the reasoning though).